Adding additional test coverage for the System.Math and System.MathF functions implemented by FCalls.

tannergooding opened the pull request:
This resolves https://github.com/dotnet/corefx/issues/12855

These tests are essentially the same as those that are implemented in C++ in the CoreCLR PAL layer: https://github.com/dotnet/coreclr/tree/master/src/pal/tests/palsuite/c_runtime

These tests cover ~~all~~ most of the interesting corner cases, except for subnormal values, and max/min domain inputs.

mellinoe commented:
I'd prefer that these be written as `[Theory]` methods, without the extra "TestData" structures, in order to match our usual testing style in corefx.

tannergooding commented:
Ah yes, I forgot I specified that in the work item. Will update.

tannergooding commented:
I've updated the `System.MathF` tests to use `Theory` and `InlineData`. I've also added a couple scenarios that were not covered (still don't have subnormal or max/min domain input tests) and added skipped `InlineData` (with bugs) for values where we aren't IEEE compliant.

I'll get around to `System.Math` tomorrow.

tannergooding commented:
I've updated the `System.Math` test now as well. There are a total of 1563 scenarios I've added.

tannergooding commented:
Looks like Ceiling on x86 when dealing with -0.xxxx values returns -0.0, while x64 returns +0.0.

The IEEE nor the C11 spec actually dicates how -0.xxxx should be handled. In either case, we should declare one behavior as 'good' and log a bug to fix the other.

tannergooding commented:
The special handling of `X = 0, and Y < 0` in the PAL layer for Mac and Linux is also incorrect, since it is not special casing odd integer values. That should be fixed as well.

tannergooding commented:
Logged:
* https://github.com/dotnet/coreclr/issues/10287 for the Math.Ceiling(-0.xxxxxxxxx) handling on x64 Windows
* https://github.com/dotnet/coreclr/issues/10288 for the Math.Floor(-0.0) handling on x86 Windows
* https://github.com/dotnet/coreclr/issues/10289 for the Math.Pow(±0, y) when y is an odd integer < 0 on non-Windows platforms

tannergooding commented:
Ok. This should be good now.

We have a ton more coverage for System.Math and System.MathF.
We have full IEEE compliance for everything except for Ceiling and Floor.
For Ceiling the tests covering inputs matching -0.xxxxxxxxx are currently disabled due to https://github.com/dotnet/coreclr/issues/10287
For the Floor(-0.0) tests, it is currently disabled due to https://github.com/dotnet/coreclr/issues/10288
We have legacy tests for the Atan2 and Pow scenarios that were changed to match IEEE compliance
We have legacy tests to match the Complex.FromPolarCoordinates scenarios that were changed to match IEEE compliance

mellinoe commented:
@tannergooding Have you run the tests against the .NET Framework? We are most likely going to need to disable many of these outside of .NET Core.

```
build.cmd -framework:netfx
msbuild <testproject> /p:TargetGroup=netfx /t:rebuildandtest
```

tannergooding commented:
@mellinoe, I haven't directly yet (doing so now).

However, the `MathF` tests are the only ones that shouldn't work on `netfx`, since they aren't available. 

The behavior between `netcore` and `netfx` should be the same for everything else, except for the `Math.Atan2` and `Math.Pow` scenarios that were just changed (I have already split these out into separate tests matching new vs legacy behavior).

tannergooding commented:
@mellinoe, there was one failing test. `Math.Tan(π/2)`, which is `16331239353195370.0` on netcore (for all platforms) and `16331778728383844.0` on netfx. I've split it out into a `Tan_PiOver2` and `Tan_PiOver2_Legacy` tests.

tannergooding commented:
Everything is green. Will merge after I get sign-off.

tannergooding commented:
*ping* for sign-off

mellinoe commented:
Hey Tanner, I've been pretty busy this week so I haven't been able to get to this. I will try to set aside some time today to review this since it's a very large change.

mellinoe commented on src/System.Runtime.Extensions/tests/System/Math.cs:
```diff
@@ -1,134 +1,654 @@
+            return (*(uint*)(&value)) == 0x00000000;
+        }
+
+        private static string ToString(double value)
```
Can you give this a more descriptive name? Maybe `ToStringPadded` ?

mellinoe commented on src/System.Runtime.Extensions/tests/System/MathF.netcoreapp.cs:
```diff
@@ -1,173 +1,613 @@
 
-        [Fact]
-        public static void Asin()
+        private static string ToString(float value)
```
Ditto about the naming

tannergooding commented on src/System.Runtime.Extensions/tests/System/Math.cs:
```diff
@@ -1,134 +1,654 @@
+
+            if (delta > variance)
+            {
+                throw new EqualException(ToString(expected), ToString(actual));
```
I didn't include it here because the xUnit EqualException doesn't have a way to easily expose the information.

However, we are essentially checking for a perfect match. That is, double can only display 15-17 digits, and we require the result to be within accurate to within 50/52 explicit mantissa bits.

tannergooding commented on src/System.Runtime.Extensions/tests/System/Math.cs:
```diff
@@ -1,134 +1,654 @@
+            }
+            else if (IsNegativeZero(actual))
+            {
+                // Do nothing, since the actual result could still be within the allowed variance
```
I should actually be checking `if (IsPositiveZero(variance) || IsNegativeZero(variance))` and throwing if true, as is done above. That way exactly precise values are handled.

tannergooding commented:
@mellinoe, I'm going to merge this as is. I have a secondary commit which addresses all the feedback that I will push up immediately after (I'd rather get this in and address the comment/naming case separately, so that it doesn't take 2 more hours to get validation on the minor changes).

mellinoe commented:
Sounds good to me. Thanks again!

karelz commented:
Thanks @tannergooding!
